p2p/discover/udp.go

```diff
@@ -587,10 +602,19 @@ func (req *ping) handle(t *udp, from *net.UDPAddr, fromID NodeID, mac []byte) er
 		ReplyTok:   mac,
 		Expiration: uint64(time.Now().Add(expiration).Unix()),
 	})
-	if !t.handleReply(fromID, pingPacket, req) {
-		// Note: we're ignoring the provided IP address right now
-		go t.bond(true, fromID, from, req.From.TCP)
-	}
+	t.handleReply(fromID, pingPacket, req)
+
+	n := NewNode(fromID, from.IP, uint16(from.Port), req.From.TCP)
+	if time.Since(t.db.lastPingReceived(fromID)) > nodeDBNodeExpiration {
+		// We haven't answered the senders ping since a long time.
+		// Verify their endpoint by pinging back. If they manage to reply
+		// their endpoint is legit and can be added to the table.
+		t.sendPing(fromID, from, func() { t.addThroughPing(n) })
+	} else {
+		// Add immediately because bond exists.
+		t.addThroughPing(n)
```

Comment:
Why does the bond exist here? We're handling the ping message here, no? Doesn't a bond mean we also got a pong?